diff -Nru --exclude *~ --exclude .* --exclude *.o --exclude include/config* linuxppc-benh-270202-nfpom/net/ipv4/netfilter/ip_nat_standalone.c linuxppc-benh-270202-gre/net/ipv4/netfilter/ip_nat_standalone.c
--- linuxppc-benh-270202-nfpom/net/ipv4/netfilter/ip_nat_standalone.c	Thu Feb 28 23:55:45 2002
+++ linuxppc-benh-270202-gre/net/ipv4/netfilter/ip_nat_standalone.c	Fri Mar  1 00:04:22 2002
@@ -330,6 +330,8 @@
 module_exit(fini);
 
 EXPORT_SYMBOL(ip_nat_setup_info);
+EXPORT_SYMBOL(ip_nat_protocol_register);
+EXPORT_SYMBOL(ip_nat_protocol_unregister);
 EXPORT_SYMBOL(ip_nat_helper_register);
 EXPORT_SYMBOL(ip_nat_helper_unregister);
 EXPORT_SYMBOL(ip_nat_expect_register);
@@ -338,4 +340,5 @@
 EXPORT_SYMBOL(ip_nat_mangle_tcp_packet);
 EXPORT_SYMBOL(ip_nat_seq_adjust);
 EXPORT_SYMBOL(ip_nat_delete_sack);
+EXPORT_SYMBOL(ip_nat_used_tuple);
 MODULE_LICENSE("GPL");
