diff -uNr linux-2.4.8-vanilla/net/ipv4/netfilter/ip_nat_helper.c linux-fixed-sack-removing/net/ipv4/netfilter/ip_nat_helper.c
--- linux-2.4.8-vanilla/net/ipv4/netfilter/ip_nat_helper.c	Sat Aug 11 15:56:27 2001
+++ linux-fixed-sack-removing/net/ipv4/netfilter/ip_nat_helper.c	Sun Aug 12 16:00:17 2001
@@ -267,8 +267,8 @@
 		/* Replace with NOPs. */
 		tcph->check
 			= ip_nat_cheat_check(*((u_int16_t *)(opt + i))^0xFFFF,
-					     0, tcph->check);
-		opt[i] = opt[i+1] = 0;
+					     (TCPOPT_NOP<<8)|TCPOPT_NOP, tcph->check);
+		opt[i] = opt[i+1] = TCPOPT_NOP;
 	}
 	else DEBUGP("Something wrong with SACK_PERM.\n");
 }

